libcrux-ml-kem/src/ind_cca/incremental.rs

```diff
@@ -0,0 +1,1003 @@
+    //<const K: usize, Vector: Operations>
+    pub trait IncrementalKeyPair {
+        /// Get the [`PublicKey1`] from this key pair as bytes.
+        ///
+        /// The output `bytes` have to be at least 64 bytes long.
+        ///
+        /// **PANICS:** if the output `bytes` are too short.
+        fn pk1_bytes(&self, bytes: &mut [u8]);
+
+        /// Get the [`PublicKey2`] from this key pair as bytes.
+        ///
+        /// The output `bytes` have to be at least K * 16 * 32 bytes long.
+        ///
+        /// **PANICS:** if the output `bytes` are too short.
+        fn pk2_bytes(&self, bytes: &mut [u8]);
+    }
+
+    impl<const K: usize, Vector: Operations> IncrementalKeyPair for MlKemKeyPairUnpacked<K, Vector> {
+        fn pk1_bytes(&self, bytes: &mut [u8]) {
+            bytes[0..32].copy_from_slice(&self.public_key().ind_cpa_public_key.seed_for_A);
+            bytes[32..64].copy_from_slice(&self.public_key().public_key_hash);
+        }
+
+        fn pk2_bytes(&self, bytes: &mut [u8]) {
+            let len = K * 16 * 32;
+            serialize_secret_key(
+                &self.public_key.ind_cpa_public_key.t_as_ntt,
+                &mut bytes[0..len],
+            )
+        }
+    }
+
+    /// The incremental public key that allows generating [`Ciphertext1`].
+    #[derive(Default)]
+    pub struct PublicKey1 {
+        pub(super) seed: [u8; 32],
+        pub(super) hash: [u8; 32],
+    }
+
+    impl TryFrom<&[u8]> for PublicKey1 {
+        type Error = Error;
+
+        fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
+            if value.len() < 64 {
+                return Err(Error::InvalidInputLength);
+            }
+
+            let mut seed = [0u8; 32];
+            seed.copy_from_slice(&value[0..32]);
+            let mut hash = [0u8; 32];
+            hash.copy_from_slice(&value[32..64]);
+            Ok(Self { seed, hash })
+        }
+    }
+
+    impl From<&[u8; 64]> for PublicKey1 {
+        fn from(value: &[u8; 64]) -> Self {
+            let mut seed = [0u8; 32];
+            seed.copy_from_slice(&value[0..32]);
+            let mut hash = [0u8; 32];
+            hash.copy_from_slice(&value[32..64]);
+            Self { seed, hash }
+        }
+    }
+
+    /// The incremental public key that allows generating [`Ciphertext2`].
+    #[repr(transparent)]
+    pub struct PublicKey2<const K: usize, Vector: Operations> {
+        pub(super) t_as_ntt: [PolynomialRingElement<Vector>; K],
+    }
+
+    /// Trait container for multiplexing over platform dependent [`MlKemKeyPairUnpacked`].
+    pub trait Keys: IncrementalKeyPair {
+        fn as_any(&self) -> &dyn Any;
+    }
+    impl<const K: usize, Vector: Operations + 'static> Keys for MlKemKeyPairUnpacked<K, Vector> {
+        fn as_any(&self) -> &dyn Any {
+            self
+        }
+    }
+
+    /// The partial ciphertext c1 - first part.
+    #[repr(transparent)]
+    pub struct Ciphertext1<const LEN: usize> {
+        pub value: [u8; LEN],
```

Comment:
it would be nice if PublicKey2 were like this, rather than just being a `&[u8]`, so it had a known, type-checked length.

Reply:
I'm having trouble understanding this... how can PK2 be platform dependent?  I'm going to pass its bytes between different platforms.  Or is it that the serialization is constant-sized, but the struct isn't?  In that case, can we make a publicly facing PublicKey2 struct that's the right size and use it for returning bytes to the user?

Reply:
That's why it needs to be serialized. Rust doesn't allow us to just cast this, that's what all the serialize functions are doing. But the size is always fixed, but depending on the platform we have bytes, or SIMD registers in the key.